game-core/src/main/java/games/strategy/engine/data/changefactory/AddUnits.java

```diff
@@ -36,11 +48,36 @@ public Change invert() {
   @Override
   protected void perform(final GameData data) {
     final UnitHolder holder = data.getUnitHolder(name, type);
-    holder.getUnitCollection().addAll(units);
+    final Collection<Unit> unitsWithCorrectOwner = getUnitsWithOwner(data);
+    holder.getUnitCollection().addAll(unitsWithCorrectOwner);
+  }
+
+  private Collection<Unit> getUnitsWithOwner(final GameData data) {
+    final Map<UUID, Unit> uuidToUnits =
+        units.stream().collect(Collectors.toMap(Unit::getId, unit -> unit));
+    return unitPlayerMap.entrySet().stream()
```

Comment:
If we're iterating over the entry set of the `unitPlayerMap`, why do we need a map for this? Can we not just iterate over the set of units and avoid the map data structure altogether?

Reply:
The original code just iterated over the set of units.  That's the problem.  We need to store the map of unit -> owner outside of the units so that when the owner changes in the unit, it doesn't change the map here.

Reply:
Cool, I think that probably should be called out to make it clear & obvious. Perhaps a javadoc comment on the map property would do the trick. WDYT?

Reply:
I added some documentation and changed the method names.  WDYT?